File: lib/faye.rb
require 'cgi'
require 'cookiejar'
require 'digest/sha1'
require 'em-http'
require 'em-http/version'
require 'eventmachine'
require 'faye/websocket'
require 'forwardable'
require 'multi_json'
require 'rack'
require 'securerandom'
require 'set'
require 'time'
require 'uri'

module Faye
  VERSION = '1.4.0'

  ROOT = File.expand_path(File.dirname(__FILE__))

  autoload :Deferrable,   File.join(ROOT, 'faye', 'mixins', 'deferrable')
  autoload :Logging,      File.join(ROOT, 'faye', 'mixins', 'logging')
  autoload :Publisher,    File.join(ROOT, 'faye', 'mixins', 'publisher')
  autoload :Timeouts,     File.join(ROOT, 'faye', 'mixins', 'timeouts')

  autoload :Namespace,    File.join(ROOT, 'faye', 'util', 'namespace')

  autoload :Engine,       File.join(ROOT, 'faye', 'engines', 'proxy')

  autoload :Channel,      File.join(ROOT, 'faye', 'protocol', 'channel')
  autoload :Client,       File.join(ROOT, 'faye', 'protocol', 'client')
  autoload :Dispatcher,   File.join(ROOT, 'faye', 'protocol', 'dispatcher')
  autoload :Scheduler,    File.join(ROOT, 'faye', 'protocol', 'scheduler')
  autoload :Extensible,   File.join(ROOT, 'faye', 'protocol', 'extensible')
  autoload :Grammar,      File.join(ROOT, 'faye', 'protocol', 'grammar')
  autoload :Publication,  File.join(ROOT, 'faye', 'protocol', 'publication')
  autoload :Server,       File.join(ROOT, 'faye', 'protocol', 'server')
  autoload :Subscription, File.join(ROOT, 'faye', 'protocol', 'subscription')

  autoload :Error,        File.join(ROOT, 'faye', 'error')
  autoload :Transport,    File.join(ROOT, 'faye', 'transport', 'transport')

  autoload :RackAdapter,  File.join(ROOT, 'faye', 'adapters', 'rack_adapter')
  autoload :StaticServer, File.join(ROOT, 'faye', 'adapters', 'static_server')

  BAYEUX_VERSION   = '1.0'
  JSONP_CALLBACK   = 'jsonpcallback'
  CONNECTION_TYPES = %w[long-polling cross-origin-long-polling callback-polling websocket eventsource in-process]

  MANDATORY_CONNECTION_TYPES = %w[long-polling callback-polling in-process]

  class << self
    attr_accessor :logger
  end

  def self.ensure_reactor_running!
    Engine.ensure_reactor_running!
  end

  def self.random(*args)
    Engine.random(*args)
  end

  def self.client_id_from_messages(messages)
    first = [messages].flatten.find { |m| m['channel'] == '/meta/connect' }
    first && first['clientId']
  end

  def self.copy_object(object)
    case object
    when Hash
      clone = {}
      object.each { |k,v| clone[k] = copy_object(v) }
      clone
    when Array
      clone = []
      object.each { |v| clone << copy_object(v) }
      clone
    else
      object
    end
  end

  def self.to_json(value)
    case value
      when Hash, Array then MultiJson.dump(value)
      when String, NilClass then value.inspect
      else value.to_s
    end
  end

  def self.async_each(list, iterator, callback)
    n       = list.size
    i       = -1
    calls   = 0
    looping = false

    loop, resume = nil, nil

    iterate = lambda do
      calls -= 1
      i += 1
      if i == n
        callback.call if callback
      else
        iterator.call(list[i], resume)
      end
    end

    loop = lambda do
      unless looping
        looping = true
        iterate.call while calls > 0
        looping = false
      end
    end

    resume = lambda do
      calls += 1
      loop.call
    end
    resume.call
  end
end

File: javascript/faye.js
'use strict';

var Faye = {
  VERSION:          '<%= Faye::VERSION %>',

  BAYEUX_VERSION:   '<%= Faye::BAYEUX_VERSION %>',
  ID_LENGTH:        <%= Faye::Engine::ID_LENGTH %>,
  JSONP_CALLBACK:   '<%= Faye::JSONP_CALLBACK %>',
  CONNECTION_TYPES: ['long-polling', 'cross-origin-long-polling', 'callback-polling', 'websocket', 'eventsource', 'in-process'],

  MANDATORY_CONNECTION_TYPES: ['long-polling', 'callback-polling', 'in-process'],

  ENV: (typeof window !== 'undefined') ? window : global,

  extend: function(dest, source, overwrite) {
    if (!source) return dest;
    for (var key in source) {
      if (!source.hasOwnProperty(key)) continue;
      if (dest.hasOwnProperty(key) && overwrite === false) continue;
      if (dest[key] !== source[key])
        dest[key] = source[key];
    }
    return dest;
  },

  random: function(bitlength) {
    bitlength = bitlength || this.ID_LENGTH;
    if (bitlength > 32) {
      var parts  = Math.ceil(bitlength / 32),
          string = '';
      while (parts--) string += this.random(32);
      var chars = string.split(''), result = '';
      while (chars.length > 0) result += chars.pop();
      return result;
    }
    var limit   = Math.pow(2, bitlength) - 1,
        maxSize = limit.toString(36).length,
        string  = Math.floor(Math.random() * limit).toString(36);

    while (string.length < maxSize) string = '0' + string;
    return string;
  },

  clientIdFromMessages: function(messages) {
    var first = [].concat(messages)[0];
    return first && first.clientId;
  },

  copyObject: function(object) {
    var clone, i, key;
    if (object instanceof Array) {
      clone = [];
      i = object.length;
      while (i--) clone[i] = Faye.copyObject(object[i]);
      return clone;
    } else if (typeof object === 'object') {
      clone = (object === null) ? null : {};
      for (key in object) clone[key] = Faye.copyObject(object[key]);
      return clone;
    } else {
      return object;
    }
  },

  commonElement: function(lista, listb) {
    for (var i = 0, n = lista.length; i < n; i++) {
      if (this.indexOf(listb, lista[i]) !== -1)
        return lista[i];
    }
    return null;
  },

  indexOf: function(list, needle) {
    if (list.indexOf) return list.indexOf(needle);

    for (var i = 0, n = list.length; i < n; i++) {
      if (list[i] === needle) return i;
    }
    return -1;
  },

  map: function(object, callback, context) {
    if (object.map) return object.map(callback, context);
    var result = [];

    if (object instanceof Array) {
      for (var i = 0, n = object.length; i < n; i++) {
        result.push(callback.call(context || null, object[i], i));
      }
    } else {
      for (var key in object) {
        if (!object.hasOwnProperty(key)) continue;
        result.push(callback.call(context || null, key, object[key]));
      }
    }
    return result;
  },

  filter: function(array, callback, context) {
    var result = [];
    for (var i = 0, n = array.length; i < n; i++) {
      if (callback.call(context || null, array[i], i))
        result.push(array[i]);
    }
    return result;
  },

  asyncEach: function(list, iterator, callback, context) {
    var n       = list.length,
        i       = -1,
        calls   = 0,
        looping = false;

    var iterate = function() {
      calls -= 1;
      i += 1;
      if (i === n) return callback && callback.call(context);
      iterator(list[i], resume);
    };

    var loop = function() {
      if (looping) return;
      looping = true;
      while (calls > 0) iterate();
      looping = false;
    };

    var resume = function() {
      calls += 1;
      loop();
    };
    resume();
  },

  // http://assanka.net/content/tech/2009/09/02/json2-js-vs-prototype/
  toJSON: function(object) {
    return JSON.stringify(object, function(key, value) {
      return (this[key] instanceof Array) ? this[key] : value;
    });
  },

  logger: function(message) {
    if (typeof console !== 'undefined') console.log(message);
  },

  timestamp: function() {
    var date   = new Date(),
        year   = date.getFullYear(),
        month  = date.getMonth() + 1,
        day    = date.getDate(),
        hour   = date.getHours(),
        minute = date.getMinutes(),
        second = date.getSeconds();

    var pad = function(n) {
      return n < 10 ? '0' + n : String(n);
    };

    return pad(year) + '-' + pad(month) + '-' + pad(day) + ' ' +
           pad(hour) + ':' + pad(minute) + ':' + pad(second);
  }
};

if (typeof module !== 'undefined')
  module.exports = Faye;
else if (typeof window !== 'undefined')
  window.Faye = Faye;

